---
output: html_document
---

```{r setup}
library(tidyverse)
library(conflicted)

# for wrangling GRanges necessary to use annotateGenes()
library(plyranges)
# annotation information for annotating the genes 
# EDIT 2019-09-05 - suggest calling libraries explicitly in code (changed below)
# because loading the library makes R VERY SLOW and you only need it once
# library(TxDb.Hsapiens.UCSC.hg38.knownGene)
# has the function annotateGenes(), which we'll use to annotate the genes
# EDIT 2019-09-05 - suggest calling libraries explicitly in code (changed below)
# because loading the library makes R VERY SLOW and you only need it once
# library(bumphunter)

filter <- dplyr::filter

knitr::opts_chunk$set(echo = TRUE)
```

## Annotate Genomic Intervals (Using Bioconductor Packages)

Instructions on a simple wasy to use Bioconductor packages to annotate genomic ranges with the closest gene.

### Wrangle Data

Fake data to use in this tutorial.

```{r}
data = tibble(chrom = c(rep('chr1', 3), rep('chr4', 3), rep('chrX', 3)),
              start = c(runif(3, 1000, 200000000), runif(3, 1000, 100000000),
                        runif(3, 1000, 100000000)),
              end = start + c(runif(9, 0, 100)))
```

Needs to be a `GRange`. The main section of the `GRanges` container **requires** the chromosome, and start and end positions, and **optionally** can have the strand and the width of the interval. If your chromsome, start, and end columns in your data are not named "seqnames", "start", and "end", you must specify which columns they are in the `as_granges()` function (also same thing if you're including width and strand). See `as_granges()` in the chunk below with seqnames as an example.

```{r}
data %>% plyranges::as_granges(seqnames = chrom) -> data_grange
```

### Wrangle the TxDb

TxDb databases are containers for storing transcript annotations. They have the syntax TxDb.[Species].[Origin].[Genome Build].[optional: Track]. Here we're using `TxDb.Hsapiens.UCSC.hg38.knownGene` because we need genes from the hg38 version of the genome. You install TxDbs through Bioconducter, ex: `Biocmanager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")` and load them as libraries. You can see `library(TxDb.Hsapiens.UCSC.hg38.knownGene)` in the setup chunk at the beginning of this document.

See [Making and Utilizing TxDb Objects](https://bioconductor.org/packages/devel/bioc/vignettes/GenomicFeatures/inst/doc/GenomicFeatures.pdf) for more information.

```{r}
# subset the Tx database for the standard chromosomes, 1 - 22, X, Y, M
# if you run just seqlevels(TxDb.Hsapiens.UCSC.hg38.knownGene)[1:25], you'll
# see it's just a vector of those chromosomes
# EDIT 2019-09-05 Stopped loading the TxDb library and called it explicitly 
# instead, because loading the library makes R very slow
txdb <- keepSeqlevels(TxDb.Hsapiens.UCSC.hg38.knownGene::
                        TxDb.Hsapiens.UCSC.hg38.knownGene,
                      seqlevels(TxDb.Hsapiens.UCSC.hg38.knownGene::
                                  TxDb.Hsapiens.UCSC.hg38.knownGene)[1:25])
```

`bumphunter`'s `annotateTranscripts()` prepares a `GRange` from a TxDb object for the `matchGenes()` function. It pulls the locations of genes from the TxDb object and annotates them with transcription start and end sites, and several different naming conventions.

```{r}
# extract the gene information for matchGenes()
# EDIT 2019-09-05 Stopped loading the bumphunter library and called it 
# explicitly instead, because it seems to slow R down and interfere with other
# packages
genes <- bumphunter::annotateTranscripts(txdb)
```

### Annotate Data with Closest Gene

`bumphunter`'s `matchGenes()` annotates a genomic region with the closest gene. First, it takes a `GRange` with genomic regions you want to annotate, then the `GRange` you prepared using `annotateTranscripts()` containing annotated genes. It returns a `data.frame` with the following columns:

- **name**: HGNC name of the closest gene
- **annotation**: RefSeq ID for the closest gene; if there are transcript variants, will include all RefSeq IDs
- **description**: factor giving the relative spatial relationship between the genomic region and the closest gene; possible levels `c("upstream", "promoter", "overlaps 5'", "inside intron", "inside exon", "covers exon(s)", "overlaps exon upstream", "overlaps exon downstream", "overlaps two exons", "overlaps 3'", "close to 3'", "downstream", "covers")`
- **region**: facter that's basically the same as description, but less detailed; possible levels `c("upstream", "promoter", "overlaps 5'", "inside", "overlaps 3'", "close to 3'", "downstream", "covers")`
- **distance**: distance between the genomic region and the 5' end/transcription start site (TSS) of the gene
- **subregion**: factor that adds more clarifying detail to the region if there, otherwise `NA`; possible levels `c("inside intron", "inside exon", "covers exon(s)", "overlaps exon upstream", "overlaps exon downstream", "overlaps two exons")` or `NA`
- **insideDistance**: if the genomic region is inside the gene, the distance past the 5' end of the gene, otherwise `NA`
- **exonnumber**: if the genomic region is inside an exon of the gene which exon it is, or `NA`
- **nexons**: number of exons in the closest gene to the genomic region
- **UTR**: if the genomic region is inside the gene's untranslated region (UTR), a factor describing the UTR, or `NA`; possible levels `c("inside transcription region", "5' UTR", "overlaps 5' UTR", "3'UTR", "overlaps 3'UTR", "covers transcription region")` or `NA`
- **strand**: which strand is the closest gene on, '+' or '-'
- **geneL**: the length of the closest gene to the genomic region
- **Entrez**: the Entrez ID of the closest gene to the genomic region
- **subjectHits**: Index in the subject, aka the TxDb GRange, of the gene

```{r}
# EDIT 2019-09-05 - decided calling the package explicitly was better for
# speed/conflicts
bumphunter::matchGenes(data_grange, genes) -> gene_anno
```







